checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt

```diff
@@ -0,0 +1,318 @@
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        assertFalse { accessCheckoutCVV.isValid() }
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state and onUpdateLengthFilter event twice - when cvv is partially valid and on focus and cardValidator is set`() {
+        givenCvvIsOnFocus()
+        givenCvvIsPartiallyValid()
+
+        createNewAccessCheckoutCVV()
+
+        verify(cardListener).onUpdate(cvvView, true)
+        verify(cardListener, times(2)).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with invalid state and onUpdateLengthFilter event twice - when cvv is invalid and on focus and cardValidator is set`() {
+        givenCvvIsOnFocus()
+        givenCvvIsInvalid()
+
+        createNewAccessCheckoutCVV()
+
+        verify(cardListener).onUpdate(cvvView, false)
+        verify(cardListener, times(2)).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state and onUpdateLengthFilter event twice - when cvv is complete and valid and on focus and cardValidator is set`() {
+        givenCvvIsOnFocus()
+        givenCvvIsCompletelyValid()
+
+        createNewAccessCheckoutCVV()
+
+        verify(cardListener).onUpdate(cvvView, true)
+        verify(cardListener, times(2)).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with invalid state and onUpdateLengthFilter event once - when cvv is partially valid and off focus and cardValidator is set`() {
+        givenCvvIsOffFocus()
+        givenCvvIsPartiallyValid()
+
+        createNewAccessCheckoutCVV()
+
+        verify(cardListener).onUpdate(cvvView, false)
+        verify(cardListener).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with invalid state and onUpdateLengthFilter event once - when cvv is invalid and off focus and cardValidator is set`() {
+        givenCvvIsOffFocus()
+        givenCvvIsInvalid()
+
+        createNewAccessCheckoutCVV()
+
+        verify(cardListener).onUpdate(cvvView, false)
+        verify(cardListener).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state and onUpdateLengthFilter event once - when cvv is complete and valid and off focus and cardValidator is set`() {
+        givenCvvIsOffFocus()
+        givenCvvIsCompletelyValid()
+
+        createNewAccessCheckoutCVV()
+
+        verify(cardListener).onUpdate(cvvView, true)
+        verify(cardListener).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state twice and onUpdateLengthFilter event thrice - when cvv is partially valid and onUpdateCVV is called`() {
+        givenCvvIsOnFocus()
+        givenCvvIsPartiallyValid()
+
+        createNewAccessCheckoutCVV().onUpdateCVV(cvvValue)
```

Comment:
Would the following help to test only the events triggered by the call to onUpdateCVV? (which is what we test here)

```
givenCvvIsOnFocus()
givenCvvIsPartiallyValid()
val  checkout =  createNewAccessCheckoutCVV()
reset(cardListener)

checkout.onUpdateCVV(cvvValue)
```

Reply:
good idea, done this now